### PR TITLE
Fix MDS shutdown deadlocks and timeouts

### PR DESCRIPTION
**Pull request by MaxKellermann**
This PR fixes several bugs that delay the MDS process shutdown by several seconds. With these changes, the MDS process exits almost instantly.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [X] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [X] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [X] No tests

### CODE REVIEW

File: src/mds/Beacon.cc [hunk 1]
```diff
@@ -73,6 +75,7 @@ void Beacon::init(const MDSMap &mdsmap)
   _notify_mdsmap(mdsmap);
 
   sender = std::thread([this]() {
+    ceph_pthread_setname(pthread_self(), "beacon");
```
Comment (batrick) [inline src/mds/Beacon.cc:78]: is this just a cleanup patch or does it have some useful side-effect?

**Review by batrick — CHANGES_REQUESTED** [review]
Pretty much every commit should be a separate PR. If appropriate for backport, each PR should have a tracker ticket:

Please create a [tracker issue](https://tracker.ceph.com/) so that this issue can be tracked for backporting.

Then, [please annotate the commit which fixes/resolves a Ceph tracker issue with](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#fixes-lines):

`Fixes: http://tracker.ceph.com/issues/...`

This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge. See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

File: src/mds/Beacon.cc (hunk 1 above)
Comment (MaxKellermann) [reply src/mds/Beacon.cc:78]: Both. The useful side effect is that top/gdb/etc. show a proper thread name. No effect for the process itself.

**Comment by MaxKellermann** [comment]
> Pretty much every commit should be a separate PR.
[...]
> This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge

Please help me understand this policy. Even if this is a single PR, there are still 4 distinct commits, and each commit can be identified to fix one thing. What does splitting into 4 PRs change for that?

**Comment by MaxKellermann** [comment]
I split this PR into 4 separate PRs, each with one commit. Not my taste, but if you prefer it that way...

**Comment by batrick** [comment]
> > Pretty much every commit should be a separate PR.
> > [...]
> > This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge
> 
> Please help me understand this policy. Even if this is a single PR, there are still 4 distinct commits, and each commit can be identified to fix one thing. What does splitting into 4 PRs change for that?

It simplifies backporting. We don't necessarily want to backport all of these changes.

**Comment by MaxKellermann** [comment]
I don't get that - you can easily backport single commits with `git cherry-pick`. I forward-port individual patches from `reef` to `main` using `stg pick` which is mostly the same as `git cherry-pick` but for stgit. (My stgit stack has grown to 250 patches. I can't submit more PRs right now due to dependencies on other unmerged PRs.)

**Pull request closed without merging**